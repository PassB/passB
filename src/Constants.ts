export enum executionContext {
  background = 'background',
  popup = 'popup',
  options = 'options',
}
