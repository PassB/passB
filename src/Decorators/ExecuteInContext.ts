// tslint:disable:no-any

import 'reflect-metadata';

export type ExecutionContext = 'background' | 'popup' | 'options' | undefined;
const executionContexts = ['background', 'popup'];

function getActionName(constructor: Function, propertyKey: string): string {
  return `${constructor.name}.${propertyKey}`;
}

interface AsynchronousRequest {
  action: string;
  params: any[];
}

export const executeInCorrectContext = () =>
  <R, T extends (this: any, ...args: any[]) => Promise<R>>
    (
      target: object,
      propertyKey: string,
      descriptor: TypedPropertyDescriptor<T>,
    ): TypedPropertyDescriptor<T> => {

    const executionContext = Reflect.getMetadata('executionContext', target, propertyKey);
    if (!executionContexts.includes(executionContext)) {
      throw new Error(`using ${executeInCorrectContext.name} `
        + 'decorator, but not declaring @Reflect.metadata("executionContext", "...") correctly',
      );
    }

    const wrappedFunction = descriptor.value || (() => Promise.reject(void 0));

    return {
      value: (async function wrapper(this: any, ...params: any[]): Promise<R> {
        if (window.executionContext === executionContext) {
          // console.debug('executing synchonously %s.%s', target.constructor.name, propertyKey);
          return await wrappedFunction.apply(this, params);
        }

        // console.debug('executing asynchonously %s.%s', target.constructor.name, propertyKey);
        return await browser.runtime.sendMessage({
          action: getActionName(target.constructor, propertyKey),
          params: [...params],
        });
      }) as T,
    };
  };

export const AsynchronousCallable = () =>
  <T extends { new(...args: any[]): {} }>(constructor: T): T =>
    class extends constructor {
      public constructor(...args: any[]) {
        super(...args);

        for (let obj = this; !!obj; obj = Object.getPrototypeOf(obj)) { // tslint:disable-line:no-this-assignment
          for (const propertyKey of Object.getOwnPropertyNames(obj)) {
            const executionContext = Reflect.getMetadata('executionContext', this, propertyKey);
            if (executionContext && window.executionContext === executionContext) {

              browser.runtime.onMessage.addListener(
                (request: AsynchronousRequest, sender: object, sendResponse: Function) => {
                  if (request.action !== getActionName(constructor, propertyKey)) {
                    return false;
                  }

                  const wrappedFn: ((...args: any[]) => Promise<any>) = (this as any)[propertyKey];
                  wrappedFn.apply(this, request.params).then(sendResponse);
                  return true;
                },
              );

            }
          }
        }
      }
    };

export const AsynchronousCallableServiceFactory =
  <X, T extends { new(...args: any[]): X }>(type: T): ((...args: any[]) => X) =>
    ((...args: any[]): X => (new (AsynchronousCallable()(type))(...args)));
