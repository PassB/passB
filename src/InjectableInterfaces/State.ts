import {Store} from 'redux';
import {Persistor} from 'redux-persist';
import {StoreContents} from '../State/State';

export interface State {
  hydrated: Promise<void>;

  getPersistor(): Persistor;

  getStore(): Store<StoreContents>;
}
