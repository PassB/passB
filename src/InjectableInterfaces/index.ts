export {Extension} from './Extension';
export {FileFormat} from './FileFormat';
export {Filler} from './Filler';
export {Matcher} from './Matcher';
export {PassB} from './PassB';
export {PassCli} from './PassCli';
export {State} from './State';
export {Strategy} from './Strategy';
export {StorageAdaper} from './StorageAdapter';
