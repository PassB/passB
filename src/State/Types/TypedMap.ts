import {fromJS, List, Map} from 'immutable';

type SimpleValues = string | number | boolean | RegExp;

type AllowedValue =
  SimpleValues |
  AllowedMap |
  AllowedList |
  TypedMap<any> | // tslint:disable-line:no-any
  undefined;

interface AllowedList extends List<AllowedValue> {
}

interface AllowedMap extends Map<string, AllowedValue> {
}

export type MapTypeAllowedData<DataType> = {
  [K in keyof DataType]: AllowedValue;
};

export interface TypedMap<DataType extends MapTypeAllowedData<DataType>> extends Map<string, AllowedValue> { // tslint:disable-line:no-any
  toJS(): DataType;

  get<K extends keyof DataType>(key: K, notSetValue?: DataType[K]): DataType[K];

  set<K extends keyof DataType>(key: K, value: DataType[K]): this;
}

export function createTypedMap<DataType extends MapTypeAllowedData<DataType>>(data: DataType): TypedMap<DataType> {
  return fromJS(data);
}
