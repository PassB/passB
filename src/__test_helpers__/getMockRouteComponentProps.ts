import {RouteComponentProps} from 'react-router';

export function getMockRouteComponentProps<P>(data: P): RouteComponentProps<P> {
  const location = {
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: {},
  };
  return {
    match: {
      isExact: true,
      params: data,
      path: '',
      url: '',
    },
    location,
    history: {
      length: 2,
      action: 'POP',
      location,
      push: jest.fn(),
      replace: jest.fn(),
      go: jest.fn(),
      goBack: jest.fn(),
      goForward: jest.fn(),
      block: jest.fn(),
      createHref: jest.fn(),
      listen: jest.fn(),
    },
    staticContext: {},
  };
}
