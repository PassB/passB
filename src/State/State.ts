import {fromJS, List, Map} from 'immutable';
import {
  applyMiddleware, combineReducers, createStore, Action, Reducer, Store,
} from 'redux';
import {persistReducer, persistStore, Persistor} from 'redux-persist';
import immutableTransform = require('redux-persist-transform-immutable');
import { composeWithDevTools } from 'remote-redux-devtools';
import {Service} from 'typedi';
import {LazyInject} from '../Decorators/LazyInject';
import {storageSyncListener, storageSyncMiddleware} from './syncMiddleware';
import {BrowserStorageAdapter, StorageAdaper} from './BrowserStorageAdapter';
import {reducer as OptionsReducer, OptionsState} from './Options';

export interface StoreContents {
  options: OptionsState;
}

export const getOptionsFromState = (state: StoreContents) => state.options;

@Service()
export class State {
  public readonly hydrated: Promise<void>;
  private store: Store<StoreContents>;
  private persistor: Persistor;

  @LazyInject(() => BrowserStorageAdapter)
  private storageAdapter: StorageAdaper;

  public constructor() {
    const reducer: Reducer<StoreContents> = persistReducer(
      {
        key: 'root',
        storage: this.storageAdapter,
        transforms: [immutableTransform({})],
      },
      combineReducers({
        options: OptionsReducer,
      }),
    );

    this.store = createStore(
      reducer,
      composeWithDevTools(
        applyMiddleware(
          storageSyncMiddleware(this.storageAdapter),
        ),
      ),
    );

    this.persistor = persistStore(this.store);
    this.hydrated = new Promise((done: () => void) => {
      let unsubscribe: () => void;
      const checkHydrated = () => {
        if (this.persistor.getState()) {
          if (unsubscribe) {
            unsubscribe();
          }
          done();
        }
      };
      unsubscribe = this.persistor.subscribe(checkHydrated);
      checkHydrated();
    });

    storageSyncListener(this.store, this.storageAdapter);
  }

  public getOptions(): OptionsState {
    return getOptionsFromState(this.store.getState());
  }

  public getPersistor(): Persistor {
    return this.persistor;
  }

  public getStore(): Store<StoreContents> {
    return this.store;
  }

  public dispatch(action: Action): void {
    this.hydrated.then(() => this.store.dispatch(action));
  }
}
