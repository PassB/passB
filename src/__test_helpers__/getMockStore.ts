import {combineReducers, createStore, Store} from 'redux';
import {StoreContents} from 'InjectableInterfaces/State';
import {initialState as initialOptionsState, reducer as OptionsReducer} from 'State/Options/Reducers';
import {initialState as initialPassEntriesState, reducer as PassEntryReducer} from 'State/PassEntries/Reducers';

const reducer = combineReducers<StoreContents>({
  options: OptionsReducer,
  passEntries: PassEntryReducer,
});

export const getMockStore = (): Store<StoreContents> => createStore<StoreContents>(
  reducer,
  {
    options: initialOptionsState,
    passEntries: initialPassEntriesState,
  },
);
