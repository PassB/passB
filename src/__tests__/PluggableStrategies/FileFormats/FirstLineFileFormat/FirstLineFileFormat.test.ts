import {Store} from 'redux';
import {getMockStore} from '__test_helpers__/getMockStore';
import {executionContext} from 'Constants';
import {container, Interfaces, Symbols} from 'Container';
import {setExecutionContext} from 'Decorators/ExecuteInContext';
import {StoreContents} from 'InjectableInterfaces/State';
import {
  FirstLineFileFormat,
  Options,
  UsernameStyle,
} from 'PluggableStrategies/FileFormats/FirstLineFileFormat/FirstLineFileFormat';
import {createTypedMap, TypedMap} from 'State/Types/TypedMap';

const passwordFileName1 = 'test/file/path';

const passwordFile1 = [
  'firstLinePw',
  'testUser',
  'login: testUser2',
  'password: specifiedPw',
];

const getOptions = (partial: Partial<Options> = {}): TypedMap<Options> => createTypedMap({
  usernameStyle: 'SecondLine' as UsernameStyle,
  ...partial,
});

describe('FirstLineFileFormat', () => {
  beforeEach(() => {
    container.snapshot();
    setExecutionContext('test');
  });

  afterEach(() => {
    container.restore();
  });

  it('dispatches expected defaultOptions in background context', () => {
    const dispatch = jest.fn();

    setExecutionContext(executionContext.background);
    container.rebind(Symbols.State).toConstantValue({getStore: () => ({dispatch})})
    container.resolve(FirstLineFileFormat);

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  it('returns the first line as password', () => {
    const fileFormat = container.resolve(FirstLineFileFormat);
    expect(fileFormat.getPassword(passwordFile1, passwordFileName1)).toBe('firstLinePw');
  });

  it('returns undefined on usernameStyle None', () => {
    const instance = container.resolve(FirstLineFileFormat);
    Object.defineProperty(instance, 'options', {value: getOptions({usernameStyle: 'None'})});
    expect(instance.getUsername(passwordFile1, passwordFileName1)).toBe(void 0);
  });

  test('returns the second line on usernameStyle SecondLine', () => {
    const instance = container.resolve(FirstLineFileFormat);
    Object.defineProperty(instance, 'options', {value: getOptions({usernameStyle: 'SecondLine'})});
    expect(instance.getUsername(passwordFile1, passwordFileName1)).toBe('testUser');
  });

  test('returns the last path part at usernameStyle LastPathPart', () => {
    const instance = container.resolve(FirstLineFileFormat);
    Object.defineProperty(instance, 'options', {value: getOptions({usernameStyle: 'LastPathPart'})});
    expect(instance.getUsername(passwordFile1, passwordFileName1)).toBe('path');
  });

});
