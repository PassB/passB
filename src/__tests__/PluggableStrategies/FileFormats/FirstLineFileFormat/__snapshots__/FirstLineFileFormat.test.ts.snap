// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstLineFileFormat dispatches expected defaultOptions in background context 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "options": Immutable.Map {
          "usernameStyle": "SecondLine",
        },
        "strategyName": "FirstLineFileFormat",
        "strategyType": "FileFormat",
      },
      "type": "OPTIONS/SET_STRATEGY_DEFAULT_OPTIONS",
    },
  ],
]
`;
